---
title: "Using PASTA for Analyzing APA in PBMCs"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using PASTA for Analyzing APA in PBMCs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
#remotes::install_github(repo = "satijalab/PASTA")
devtools::load_all()
library(PASTA)
```

## Map Azimuth to RNA
First, you can quantify RNA and map to azimuth in the standard way. 
```{r read_in_seurat_object}
pbmc <- readRDS(file = "/brahms/shared/PASTA/pbmc.rds")
projected.umap <- readRDS('/brahms/shared/PASTA/umap_pbmcs.Rds')
pbmc <- pbmc[, Cells(projected.umap)]
pbmc[['umap']] <- projected.umap
predictions <- read.delim('/brahms/shared/PASTA/azimuth_pred_pbmcs.tsv', row.names = 1)
pbmc <- AddMetaData(
  object = pbmc,
  metadata = predictions)
DimPlot(pbmc, group.by = 'predicted.celltype.l2', label = T) + NoLegend()
```

## Make polyA assay

```{r load_polyA}
pA.count.file <- "/brahms/shared/PASTA/pA_counts.tab.gz"
peak.file <- "/brahms/shared/PASTA/pA_polyA_peaks.gff"
fragment.file <- "/brahms/shared/PASTA/10k_PBMC.blocks.sort.uniq.bed.gz"

pA.counts = ReadPolyApipe( counts.file = pA.count.file, 
                           peaks.file = peak.file, 
                           filter.chromosomes = TRUE,
                           min.features = 10, 
                           min.cells = 25)

#add gemgroup information to match cellranger output
colnames(pA.counts) <- paste0( colnames(pA.counts), '-1')

# create polyAsite.Assay
pA.assay = CreatePolyAsiteAssay(counts = pA.counts , genome = "hg38", fragments = fragment.file, validate.fragments = FALSE)
```

## Add polyA site assay to seurat object
```{r, include=FALSE}
# Add polyAsite.Assay to object
cells = intersect( Cells(pbmc) , Cells(pA.assay))
pbmc = subset(pbmc, cells = cells)
pA.assay = as( subset(pA.assay, cells = cells) , "polyAsiteAssay")
pbmc[['pA']] <- pA.assay

#for now read 

############################################
# Annotate Peaks
############################################

# Add annotation
require("plyranges")
require(GenomicFeatures)
# Annotate from EnsDB is done similarly to polyApipe
# In addition, we annotate tandem UTRs
# S.sub = AnnotatePASfromEnsDB(object = S.sub, assay = "pA", species = "Homo sapiens", version = 100, hard_extension = 20, extension = 2000)
# Annotate from GTF is motivated by and done similar to the Sierra package
gtf.file = '~/data/CPAcore/Sierra/genes.gtf'
BSgenome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38

```


